dosbox-staging/dosbox-staging
# Unicode engine improvements

# Description

This PR brings several small Unicode changes:

1. Added a specialized routines to UTF-8 <-> 437 conversion for the future Unicode support in the filesystem. The 437 -> UTF-8 one is just a syntactic sugar, but UTF-8 -> 437 is simplified and does not handle combining marks - so, for example, for character 0x94 it requires precisely `U+00F6`; `U+006F` followed by `U+0308` (which, from the host OS filesystem perspective differs from `U+00F6`) won't be accepted.

2. There was a minor mistake in code page 1118 definition, this is now fixed (but the character was similar).

3. The engine used to have problems if DOS code point was a mere combining mark (not a regular character + combining mark), it did not even accept such definition.
I have worked this around by silently replacing such definition with SPACE + combining mark, so that now I could add the list of combining marks for all the supported Unicode blocks.
Knowing that some code point is a combining mark is important for the engine, as it strips the combining marks when converting UTF-8 text to DOS if the original grapheme can't be represented in the given code page - i. e. the original UTF-8 grapheme `l̃` can be represented as letter `l` after conversion to DOS.
Currently this only affects one code for which we do provide a screen font (864, Arabic) - but there are few more cases like that in code pages where we do not bundle screen fonts yet.

4. After some testing of the result of point 3. and further investigation, I have added an entry to override the official Unicode mapping, so that we now use an isolated form of the _Arabic Shadda_ instead of the combining mark (probably the character wasn't available in Unicode when the original mapping was created).


## Related issues

https://github.com/dosbox-staging/dosbox-staging/issues/2619 (tor the extended Unicode API)


# Release notes

(we can mention that our Unicode engine has seen some minor tweaks and improvements, but I wouldn't go into the details)


# Manual testing

1. To test the _Arabic Shadda_ character available at code point `0xf1` of code page `864`, execute inside the DOSBox:
`keyb us 864`
`echo <with ALT pressed type 241 on the numeric keypad> | clip`
and paste the character to the host text editor.

2. To test that all code pages can still be set, execute inside the DOSBox the following BAT script and observe the logs.
[TESTCP.BAT.TXT](https://github.com/user-attachments/files/18223328/TESTCP.BAT.TXT)
Currently it prints out warnings regarding code pages _58335_ and _60258_, but these are not related to this change - the _58335_ really has the same characters at points `0x81` and `0x97` (you can set by executing `keyb us 58335` and check with some ASCII table software, for example from _DOS Navigator_ menu select _Utilities_ -> _ASCII Table_), and the _60258_ is indeed strange - just try to type uppercase letter `I` with it :D

3. Unfortunately, Unicode engine does not function under unit tests (one more thing on my personal TODO list of issues to fix), so I tested it by temporarily adding some test code to one of the commands:

```C++
#include "unicode.h"

	...

	const std::string test1 = std::string("\n\n") +
		fs_utf8_to_dos_437("ÇüéâäàåçêëèïîìÄÅ") + "\n" +
		fs_utf8_to_dos_437("ÉæÆôöòûùÿÖÜ¢£¥₧ƒ") + "\n" +
		fs_utf8_to_dos_437("áíóúñÑªº¿⌐¬½¼¡«»") + "\n" +
		fs_utf8_to_dos_437("░▒▓│┤╡╢╖╕╣║╗╝╜╛┐") + "\n" +
	        fs_utf8_to_dos_437("└┴┬├─┼╞╟╚╔╩╦╠═╬╧") + "\n" +
	        fs_utf8_to_dos_437("╨╤╥╙╘╒╓╫╪┘┌█▄▌▐▀") + "\n" +
	        fs_utf8_to_dos_437("αßΓπΣσµτΦΘΩδ∞φε∩") + "\n" +
	        fs_utf8_to_dos_437("≡±≥≤⌠⌡÷≈°∙·√ⁿ²■ ") + "\n" +
	        "\n";
	WriteOut(test1.c_str());

	const std::string test2 = std::string("\n") +
	        fs_utf8_to_dos_437("abc def ghi jkl mno pqr stu vwx yz ABC 123 ''") + "\n" +
	        fs_utf8_to_dos_437("This line should not be visible due to '\n'") + "\n" +
	        fs_utf8_to_dos_437("This line should not be visible due to 'Ł'") + "\n" +
	        fs_utf8_to_dos_437("This line should not be visible due to 'l̃'") + "\n" +
	        "\n";
	WriteOut(test2.c_str());
```

The change has been manually tested on:

- [x] Linux (trivial changes)


# Checklist

I have:
